worlds/mlss/Client.py

```diff
@@ -0,0 +1,266 @@
+                b = await bizhawk.read(ctx.bizhawk_ctx, [(0x3057, 1, "EWRAM")])
+                await asyncio.sleep(.05)
+                if b[0][0] == 0:
+                    await bizhawk.write(ctx.bizhawk_ctx, [(0x3057, [id_to_RAM(item_data.itemID)], "EWRAM"), (0x4808, [(i + 1) // 0x100, (i + 1) % 0x100], "EWRAM")])
+                else:
+                    break '''
+
+            if shopping & 0x3 == 0x3:
+                if locs_to_send != self.local_checked_locations:
+                    self.local_checked_locations = locs_to_send
+
+                    if locs_to_send is not None:
+                        await ctx.send_msgs([{
+                            "cmd": "LocationChecks",
+                            "locations": list(locs_to_send)
+                        }])
+                return
+
+            # Checking flags that aren't digspots or blocks
+            for item in nonBlock:
+                address, mask, location = item
+                if location in self.local_checked_locations:
+                    continue
+                flag_bytes = await bizhawk.read(ctx.bizhawk_ctx, [(address, 1, "EWRAM"), (0x4800, 6, "EWRAM")])
+                flag_byte = flag_bytes[0][0]
+                backup_logo = bytes([byte for byte in flag_bytes[1] if byte < 0x70]).decode("UTF-8")
+                if backup_logo != "MLSSAP":
+                    return
+                if flag_byte & mask != 0:
+                    if location in roomException:
+                        if roomException[location] != room:
+                            if (location == 0xDA0001 or location == 0x2578e7) and room == 0x79:
+                                exception = False
+                            else:
+                                exception = True
+                        else:
+                            exception = False
+                    else:
+                        exception = True
+
+                    if location in eReward:
+                        if location not in self.eUsed:
+                            self.eUsed += [location]
+                            location = eReward[len(self.eUsed) - 1]
+                        else:
+                            continue
+                    if (location in ctx.server_locations) and exception:
+                        locs_to_send.add(location)
+
+            # Check for set location flags.
+            for byte_i, byte in enumerate(bytearray(flags)):
+                for j in range(8):
+                    if j in self.checked_flags[byte_i]:
+                        continue
+                    and_value = 1 << j
+                    if byte & and_value != 0:
+                        flag_id = byte_i * 8 + (j + 1)
+                        room, item = find_key(roomCount, flag_id)
+                        pointer_arr = await bizhawk.read(ctx.bizhawk_ctx,
+                                                         [(ROOM_ARRAY_POINTER + ((room - 1) * 4), 4, "ROM")])
```

Comment:
This seems like a lot. Won't it do this `read` for every set flag in 59 bytes except for those that correspond to locations that were already checked? Are a majority of those flags locations? If there are, like, 100 flags in here used for story progress that get set over the course of a run, it looks like they'll be checked like this over and over.

Reply:
Every single bit in those 59 bytes is a flag that is in the randomizer, they are for every block and digspot. There are only a handful of bits in those 59 bytes that do not have a location tied to them.

Reply:
Got it. Emerald is much more sparse by comparison, lol.